src/plugins/list/list.c

```diff
@@ -311,6 +317,33 @@ static int runPlugins (KeySet * pluginKS, KeySet * modules, KeySet * plugins, Ke
 				keyDel (slaveKey);
 			}
 		}
+		if (ioBinding)
```

Comment:
Is this coupling necessary? (What about callbacks or some other more generic mechanism) If it is necessary: The function is already very long, can you refactor this a bit?

Reply:
It was the only solution I could find: Without a KDB handle in the Plugin structure the plugins cannot get the notification callback or I/O binding from stored in the KDB handle. So the notification library and `elektraIoSetBinding()` need to "push" callbacks to the plugins using exported functions (`open/closeNotification` and `setIoBinding`).

Since the list plugin loads its plugins internally I cannot get these functions from `system/elektra/modules/...`. Therefore the list plugin needs to act as a proxy for these exported functions. 
The plugin cannot set it directly because it lazy-loads the plugins later at the first `kdbGet()` or `kdbSet()` which is why I had to add it here.

But of course in the long-term it is not feasible to implement every possible exported function that needs to be passed to plugins. How about these two solutions for more generic exported functions?

A simple generic solution would be to have a `getNestedPlugins` function exported by the list plugin and similar plugins. The list plugin would have to load all plugins when this funciton is called for `setIoBinding` and `openNotification` to work since these are called before the first `kdbGet/Set()`.

Another solution that would make the exported functions also generic for the caller would be to have a `getInterfaces` function exported by compatible plugins. The returned "interfaces" could be a `KeySet` with well-known interface name like "notification" or "ioBinding" and the required functions (e.g. `open`, `close` for "notification", or `set` for "ioBinding").
Normal plugins would simply return their interfaces. Plugins like the list plugin could collect interfaces from its nested plugins (again after loading them) and return a KeySet that combines all interfaces from the nested plugins.

Here is an example interface-`KeySet` from a list plugin with two nested plugins A and B both implementing the "notification" interface. Plugin B also has the "ioBinding" interface:
- `/notification/#0/open` => address of `openNotification` of plugin A
- `/notification/#0/close` => address of `closeNotification` of plugin A
- `/notification/#1/open` => address of `openNotification` of plugin B
- `/notification/#1/close` => address of `closeNotification` of plugin B
- `/ioBinding/#0/set` => address of `setIoBinding` of plugin B

Both solutions would only circumvent the list plugin's lazy-loading when the notification features are used by an application. The second solution might be a bit over-engineered :)

Reply:
> need to "push" callbacks to the plugins using exported functions (open/closeNotification and setIoBinding)

There should be nothing wrong with that but the callback interface should be more generic (not specific to notification).

> The list plugin would have to load all plugins

The idea of the list plugin is to avoid exactly this. So not circumventing lazy-loading (which is the main reason for the list plugin in the first place) would be preferable.

Reply:
Ok, to sum up the requirements:

1. the callback interface on the list plugin should be generic (independent from notification & I/O binding)
2. the list plugin should not load all plugins before they are actually required within kdbGet()/kdbSet()
3. there is this notification initialization sequence: 
    - `kdbOpen(kdb, ...)`
    - `elektraIoSetBinding(kdb, binding)`: plugins are checked for `setIoBinding()` exports
    - `elektraNotificationOpen(kdb)`: plugins are checked for `openNotification()` exports
    - `kdbGet(kdb, ...)`: list plugin loads it's plugins

I can think of some "delayCallback" function exported by the list plugin which is used if either exported function is not present. It takes a function name and a list of arguments. When the list plugin initializes it's plugins it would check for the exported functions and call this function with the arguments. Since C is statically typed that will change the definition of both `setIoBinding()` and `openNotification()` to a KeySet (or any list) containing the arguments. 

Would that be better?

Reply:
Another requirement is that there should be an interface which is not entangled with details of the list plugin. The list plugin might be replaced and we want to avoid that your I/O code needs changes in such a situation.

> there is this notification initialization sequence: 

Do all these calls happen from the application?

Why not have an elektraNotificationOpen which also takes the binding as second parameter?

> Would that be better?

Yes, having a way to call functions within plugins
1.) without being entangled with list plugin details,
2.) for anyone (and not only I/O bindings)
3.) and still preserving lazyness
would be important.

This is needed for everything in the direction of #868

It would be great to have written everything together in a design decision. The discussion is already quite distributed.

Reply:
> Why not have an elektraNotificationOpen which also takes the binding as second parameter?

As long as I/O binding and notification are only used together this makes sense - in the long term I/O bindings might be used alone.

> It would be great to have written everything together in a design decision. The discussion is already quite distributed.

Okay I'll write up a design decision and add it to this PR.

Reply:
I added the decision and called it [deferred plugin calls](https://github.com/ElektraInitiative/libelektra/pull/1825/files#diff-4192311c812b52df60a9bd3eca54c457).

Reply:
I also added an initial implementation of the proposed solution. The generic functions for handling the deferred calls from the decision are now in the "list" plugin.

What would be the best place to put them? src/libs/elektra/plugin.c (with according prefix) or is an object library better which can be included on demand?

Reply:
As written as comment, plugin.c should stay minimal. Only use it for functions which manipulate the plugin structure.

For the convenience functions elektra-invoke might be suitable?